crypto/rand/drbg_rand.c

```diff
@@ -0,0 +1,929 @@
+static int drbg_hmac_generate(DRBG_CTX *dctx,
+                              unsigned char *out, size_t outlen,
+                              const unsigned char *adin, size_t adin_len)
+{
+    DRBG_HMAC_CTX *hmac = &dctx->d.hmac;
+    HMAC_CTX *hctx = hmac->hctx;
+    const unsigned char *Vtmp = hmac->V;
+
+    if (adin_len
+            && !drbg_hmac_update(dctx, adin, adin_len, NULL, 0, NULL, 0))
+        return 0;
+    for ( ; ; ) {
+        if (!HMAC_Init_ex(hctx, hmac->K, dctx->blocklength, hmac->md, NULL)
+                || !HMAC_Update(hctx, Vtmp, dctx->blocklength))
+            return 0;
+        if (!dctx->lb_valid) {
+            if (!HMAC_Final(hctx, dctx->lb, NULL))
+                return 0;
+            dctx->lb_valid = 1;
+            Vtmp = dctx->lb;
+            continue;
+        }
+        if (outlen > dctx->blocklength) {
+            if (!HMAC_Final(hctx, out, NULL))
+                return 0;
+            if (!drbg_cprng_test(dctx, out))
+                return 0;
+            Vtmp = out;
+        } else {
+            if (!HMAC_Final(hctx, hmac->V, NULL))
+                return 0;
+            if (!drbg_cprng_test(dctx, hmac->V))
+                return 0;
+            memcpy(out, hmac->V, outlen);
+            break;
+        }
+        out += dctx->blocklength;
+        outlen -= dctx->blocklength;
+    }
+    if (!drbg_hmac_update(dctx, adin, adin_len, NULL, 0, NULL, 0))
+        return 0;
+    return 1;
+}
+
+static int drbg_hmac_uninstantiate(DRBG_CTX *dctx)
+{
+    HMAC_CTX_free(dctx->d.hmac.hctx);
+    OPENSSL_cleanse(&dctx->d.hmac, sizeof(dctx->d.hmac));
+    return 1;
+}
+
+int drbg_hmac_init(DRBG_CTX *dctx)
+{
+    const EVP_MD *md = NULL;
+    DRBG_HMAC_CTX *hctx = &dctx->d.hmac;
+
+    dctx->strength = 256;
+    switch (dctx->type) {
+    default:
+        dctx->strength = 0;
+        return -2;
+    case NID_hmacWithSHA1:
+        md = EVP_sha1();
+        dctx->strength = 128;
+        break;
+    case NID_hmacWithSHA224:
+        md = EVP_sha224();
+        dctx->strength = 192;
+        break;
+    case NID_hmacWithSHA256:
+        md = EVP_sha256();
+        break;
+    case NID_hmacWithSHA384:
+        md = EVP_sha384();
+        break;
+    case NID_hmacWithSHA512:
+        md = EVP_sha512();
+        break;
+    }
+
+    dctx->instantiate = drbg_hmac_instantiate;
+    dctx->reseed = drbg_hmac_reseed;
+    dctx->generate = drbg_hmac_generate;
+    dctx->uninstantiate = drbg_hmac_uninstantiate;
+    hctx->hctx = HMAC_CTX_new();
+    hctx->md = md;
+    dctx->blocklength = EVP_MD_size(md);
+    dctx->seedlen = EVP_MD_size(md);
+    dctx->min_entropy = dctx->strength / 8;
+    dctx->max_entropy = DRBG_MAX_LENGTH;
+    dctx->min_nonce = dctx->min_entropy / 2;
+    dctx->max_nonce = DRBG_MAX_LENGTH;
+    dctx->max_pers = DRBG_MAX_LENGTH;
+    dctx->max_adin = DRBG_MAX_LENGTH;
+    dctx->max_request = 1 << 16;
+    dctx->reseed_interval = 1 << 24;
+    return 1;
+}
+
+DRBG_CTX *RAND_DRBG_get_default(void)
```

Comment:
### Default DRBG should be initialized with sensible entropy callbacks

The original FIPS 2.0 DRBG is somewhat incomplete because it implements only the deterministic part and relies on the caller to provide entropy callbacks during FIPS DRBG initialisation. These callbacks were provided by the OpenSSL 1.0.2 crypto library during the FIPS DRBG initialization (see the end of [rand_lib.c](https://github.com/openssl/openssl/blob/OpenSSL_1_0_2-stable/crypto/rand/rand_lib.c))

This code has been removed from master during the FIPS code cleanup in 700b4a4ae7c2a89ca99bfe64baef1eabfa316136. As a consequence, the DRBG_CTX currently returned by RAND_DRBG_get_default() (hence also the RAND_METHOD returned by the RAND_drbg()) is unusable unless the user (i.e., the programmer) provides entropy callbacks.

I think it would be a good idea if do_ossl_drbg_init() would take care of providing sensible defaults. You could take the removed code from 700b4a4ae7c2a89ca99bfe64baef1eabfa316136 as a starting point.